runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -369,7 +369,10 @@ public State call() throws Exception {
         } catch (IOException e) {
           State state = getState();
           if (state.isTerminal()) {
-            LOG.warn("Job is already terminated. State is {}", state);
+            LOG.warn("Cancel failed because job is already terminated. State is {}", state);
+            return state;
+          } else if (e.getMessage().contains("has terminated")) {
```

Comment:
is string matching the worst thing ever? yes, except that this is I think the only reliable signal.

We'll get to delete this code fairly soon, I think.